### PR TITLE
Adds Layers and Docks

### PR DESCRIPTION
**Pull request by willmcgugan**
Replaces the dock layout with a simpler dock rule.

The "dock" rule attribute accepts "top", "right", "bottom", and "left", and will move the widget to that edge, and exclude it from scrolling. Similar to an absolutely positioned div in HTML.

Additionally, this also implements *layers*. A container widget specified a list of layers, such as "base menu modal" to define a gross ordering for its children. Widgets specify a "layer" (singular) attribute to specify which layer a widget will go in to.

### CODE REVIEW

File: src/textual/_layout.py [hunk 1]
```diff
@@ -25,6 +27,7 @@ class WidgetPlacement(NamedTuple):
     region: Region
     widget: Widget | None = None  # A widget of None means empty space
     order: int = 0
+    fixed: bool = False
```
Comment (willmcgugan) [inline src/textual/_layout.py:30]: Sort of equivalent to `position: fixed` in CSS. Basically excludes a widget from scrolling.

File: src/textual/_styles_cache.py [hunk 2]
```diff
@@ -246,7 +246,9 @@ def post(segments: Iterable[Segment]) -> list[Segment]:
         line: Iterable[Segment]
         # Draw top or bottom borders (A)
         if (border_top and y == 0) or (border_bottom and y == height - 1):
-            border_color = border_top_color if y == 0 else border_bottom_color
+            border_color = background + (
```
Comment (willmcgugan) [inline src/textual/_styles_cache.py:249]: Adding the background allows border colors to respect alpha.

File: src/textual/color.py [hunk 3]
```diff
@@ -161,9 +165,10 @@ def clamped(self) -> Color:
     @property
     def rich_color(self) -> RichColor:
         """This color encoded in Rich's Color class."""
-        # TODO: This isn't cheap as I'd like - cache in a LRUCache ?
         r, g, b, _a = self
-        return RichColor.from_rgb(r, g, b)
+        return RichColor(
```
Comment (willmcgugan) [inline src/textual/color.py:169]: Essentially doing the same, but cutting out a few intermediate steps.

File: e2e_tests/test_apps/basic.py [hunk 4]
```diff
@@ -83,71 +88,112 @@ def render(self) -> Text:
...
-            sidebar=Widget(
-                Widget(classes="title"),
-                Widget(classes="user"),
-                OptionItem(),
-                OptionItem(),
-                OptionItem(),
-                Widget(classes="content"),
+            id="header",
+        )
+        yield from (
```
Comment (willmcgugan) [inline e2e_tests/test_apps/basic.py:111]: Quite like this way of adding a bunch of widgets.

**Review by darrenburns — APPROVED** [review]

**Pull request merged**